src/app/fdctl/run/tiles/fd_replay.c

```diff
@@ -0,0 +1,731 @@
+      // Copy over latest blockhash to slot_bank poh for updating the sysvars
+      fd_memcpy( fork->slot_ctx.slot_bank.poh.uc, ctx->blockhash.uc, sizeof(fd_hash_t) );
+      fd_block_info_t block_info[1];
+      block_info->signature_cnt = fork->slot_ctx.signature_cnt;
+      int res = fd_runtime_block_execute_finalize_tpool( &fork->slot_ctx, NULL, block_info, NULL, 1UL );
+      if( res != FD_RUNTIME_EXECUTE_SUCCESS ) {
+        FD_LOG_WARNING(("block finalize failed"));
+        *opt_filter = 1;
+        return;
+      }
+      
+      fd_blockstore_start_write( ctx->replay->blockstore );
+
+      fd_block_t * block_ = fd_blockstore_block_query( ctx->replay->blockstore, ctx->curr_slot );
+      if( FD_LIKELY( block_ ) ) {
+        block_->flags = fd_uchar_set_bit( block_->flags, FD_BLOCK_FLAG_PROCESSED );
+        memcpy( &block_->bank_hash, &fork->slot_ctx.slot_bank.banks_hash, sizeof( fd_hash_t ) );
+      }
+
+      fd_blockstore_end_write( ctx->replay->blockstore );
+
+      // Remove slot ctx from frontier once block is finalized
+      fd_fork_t * child = fd_fork_frontier_ele_remove( ctx->replay->forks->frontier, &fork->slot, NULL, ctx->replay->forks->pool );
+      child->slot = ctx->curr_slot;
+      if( FD_UNLIKELY( fd_fork_frontier_ele_query(
+          ctx->replay->forks->frontier, &ctx->curr_slot, NULL, ctx->replay->forks->pool ) ) ) {
+        FD_LOG_ERR( ( "invariant violation: child slot %lu was already in the frontier", ctx->curr_slot ) );
+      }
+      fd_fork_frontier_ele_insert( ctx->replay->forks->frontier, child, ctx->replay->forks->pool );
+
+      /* Prepare bank for next execution. */
+
+      child->slot_ctx.slot_bank.slot           = ctx->curr_slot;
+      child->slot_ctx.slot_bank.collected_fees = 0;
+      child->slot_ctx.slot_bank.collected_rent = 0;
+
+      fd_hash_t const * bank_hash = &child->slot_ctx.slot_bank.banks_hash;
+
+      fd_bank_hash_cmp_t * bank_hash_cmp = fd_exec_epoch_ctx_bank_hash_cmp( child->slot_ctx.epoch_ctx );
+      fd_bank_hash_cmp_lock( bank_hash_cmp );
+      fd_bank_hash_cmp_insert( bank_hash_cmp, ctx->curr_slot, bank_hash, 1 );
+
+      /* Try to move the bank hash comparison window forward */
+      while (1) {
+        ulong *children, nchildren, parent_slot = bank_hash_cmp->slot;
+        if ( fd_blockstore_next_slot_query( ctx->replay->blockstore, parent_slot, &children, &nchildren ) == FD_BLOCKSTORE_OK ) {
+          for (ulong i = 0; i < nchildren; i++) {
+            if( FD_LIKELY( fd_bank_hash_cmp_check( bank_hash_cmp, children[i] ) ) ) {
+              bank_hash_cmp->slot = children[i];
+              break;
+            }
+          }
+        } else {
+          FD_LOG_WARNING( ("failed at getting children of slot %lu", parent_slot) );
+        }
+        if(bank_hash_cmp->slot == parent_slot) break;
+      }
+      fd_bank_hash_cmp_unlock( bank_hash_cmp );
+    }
+  } FD_SCRATCH_SCOPE_END;
+}
+
+static void
+read_genesis( void * _ctx,
```

Comment:
Been thinking about this, can we move this genesis stuff out somewhere and dedup with runtime (at least at some point)? Really unclean to look at in here

Reply:
Can we defer to later?

Reply:
Gonna defer to later